js/ui/calendar/ui.calendar.selection.strategy.js

```diff
@@ -0,0 +1,227 @@
+        const values = [...this.dateOption('values')];
+        const alreadySelectedIndex = values.findIndex(date => date?.toDateString() === selectedValue.toDateString());
+
+        if(alreadySelectedIndex > -1) {
+            values.splice(alreadySelectedIndex, 1);
+        } else {
+            values.push(selectedValue);
+        }
+
+        this._updateCurrentDate(selectedValue);
+        this._currentDateChanged = true;
+        this.dateValue(values, e);
+    },
+
+    updateAriaSelected: function(value, previousValue) {
+        value = value ?? this.dateOption('values');
+        previousValue = previousValue ?? [];
+
+        this.callBase(value, previousValue);
+    },
+
+    getDefaultCurrentDate: function() {
+        return this._getLowestDateInArray(this.dateOption('values').filter(value => value));
+    },
+});
+
+const CalendarRangeSelectionStrategy = CalendarSelectionStrategy.inherit({
+    NAME: 'RangeSelection',
+
+    getViewOptions: function() {
+        const value = this._getValues();
+        const range = this._getDaysInRange(value[0], value[1]);
+
+        return {
+            value,
+            range,
+            selectionMode: 'range',
+            onCellHover: this._cellHoverHandler.bind(this),
+        };
+    },
+
+    selectValue: function(selectedValue, e) {
+        const [startDate, endDate] = this._getValues();
+
+        this._updateCurrentDate(selectedValue);
+        this._currentDateChanged = true;
+        if(!startDate || endDate) {
+            this.dateValue([selectedValue, null], e);
+        } else {
+            this.dateValue(startDate < selectedValue ? [startDate, selectedValue] : [selectedValue, startDate], e);
+        }
+    },
+
+    updateAriaSelected: function(value, previousValue) {
+        value = value ?? this._getValues();
+        previousValue = previousValue ?? [];
+
+        this.callBase(value, previousValue);
+    },
+
+    processValueChanged: function(value, previousValue) {
+        this.callBase(value, previousValue);
+
+        const range = this._getRange();
+        this._updateViewsOption('range', range);
+    },
+
+    getDefaultCurrentDate: function() {
+        return this._getLowestDateInArray(this._getValues().filter(value => value));
+    },
+
+    _getValues: function() {
+        const values = this.dateOption('values');
+
+        if(!values.length) {
+            return values;
+        }
+
+        let [startDate, endDate] = values;
+
+        if(startDate && endDate && startDate > endDate) {
+            [startDate, endDate] = [endDate, startDate];
+        }
+
+        return [startDate, endDate];
+    },
+
+    _getRange: function() {
+        const [startDate, endDate] = this._getValues();
+        return this._getDaysInRange(startDate, endDate);
+    },
+
+    _getDaysInRange: function(startDate, endDate) {
+        return startDate && endDate
+            ? [...dateUtils.getDatesOfInterval(startDate, endDate, DAY_INTERVAL), endDate]
```

Comment:
Cannot we implement a range strategy without settings all range days as a value? I think it's the main reason of a performance issue

Reply:
It's clear that's is the main reason, but I'm not sure how to do it in another way. I'll try couple ideas

Reply:
Now range is reduced, it can not be +-days from currentDate